### PR TITLE
Remove bad_variant_access

### PR DESCRIPTION
**Pull request by PetrovKP**

### CODE REVIEW

**Comment by PetrovKP** [comment]
/intelci: run
set daal4py_branch=dev/onedal4py
set daal4py_repo=git@github.com:PetrovKP/daal4py.git

**Comment by PetrovKP** [comment]
/intelci: restart
set daal4py_branch=dev/onedal4py
set daal4py_repo=git@github.com:PetrovKP/daal4py.git

**Comment by PetrovKP** [comment]
eb9a6d8b-1bcd-f1ae-90aa-8cdcd4b723a1

**Comment by PetrovKP** [comment]
/intelci: run
set daal4py_branch=dev/onedal4py
set daal4py_repo=git@github.com:PetrovKP/daal4py.git

File: cpp/oneapi/dal/detail/array_impl.hpp [hunk 1]
```diff
@@ -88,18 +88,17 @@ class array_impl : public base {
         if (const auto& mut_ptr = std::get_if<shared>(&data_owned_)) {
             return mut_ptr->get();
         }
-        else {
-            const auto& immut_ptr = std::get<cshared>(data_owned_);
-            return immut_ptr.get();
+        else if (const auto& immut_ptr = std::get_if<cshared>(&data_owned_)) {
+            return immut_ptr->get();
         }
+        return nullptr;
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/detail/array_impl.hpp:94]: Can you please place a `ONEDAL_ASSERT` before this `return` statement? It should never be reachable.

File: cpp/oneapi/dal/detail/array_impl.hpp [hunk 2]
```diff
@@ -163,22 +162,47 @@ class array_impl : public base {
     template <typename Y>
     void reset(const array_impl<Y>& ref, T* data, std::int64_t count) {
         if (ref.has_mutable_data()) {
-            data_owned_ = shared(std::get<1>(ref.data_owned_), data);
+            if (const auto& ptr = std::get_if<1>(&ref.data_owned_)) {
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/detail/array_impl.hpp:165]:
```suggestion
            if (const auto& ptr = std::get_if<shared>(&ref.data_owned_)) {
```
For consistency

File: cpp/oneapi/dal/detail/array_impl.hpp [hunk 3]
```diff
@@ -163,22 +162,47 @@ class array_impl : public base {
...
+                data_owned_ = shared(*ptr, data);
+            }
+            else {
+                throw internal_error(
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/detail/array_impl.hpp:169]: Replace by assert - this branch shall be unreachable

File: cpp/oneapi/dal/detail/array_impl.hpp [hunk 4]
```diff
@@ -163,22 +162,47 @@ class array_impl : public base {
...
+                    dal::detail::error_messages::array_does_not_contain_mutable_data());
+            }
         }
         else {
-            data_owned_ = shared(std::get<0>(ref.data_owned_), data);
+            if (const auto& ptr = std::get_if<0>(&ref.data_owned_)) {
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/detail/array_impl.hpp:174]:
```suggestion
            if (const auto& ptr = std::get_if<cshared>(&ref.data_owned_)) {
```

File: cpp/oneapi/dal/detail/array_impl.hpp [hunk 5]
```diff
@@ -163,22 +162,47 @@ class array_impl : public base {
...
+                data_owned_ = shared(*ptr, data);
+            }
+            else {
+                throw internal_error(
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/detail/array_impl.hpp:178]: I'd suggest changing it to a more meaningful error message that the reference array is empty and does not contain ownership structure.

File: cpp/oneapi/dal/detail/array_impl.hpp [hunk 6]
```diff
@@ -163,22 +162,47 @@ class array_impl : public base {
...
+
         count_ = count;
         reset_policy(ref);
     }
 
     template <typename Y>
     void reset(const array_impl<Y>& ref, const T* data, std::int64_t count) {
         if (ref.has_mutable_data()) {
-            data_owned_ = cshared(std::get<1>(ref.data_owned_), data);
+            if (const auto& ptr = std::get_if<1>(&ref.data_owned_)) {
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/detail/array_impl.hpp:190]:
```suggestion
            if (const auto& ptr = std::get_if<shared>(&ref.data_owned_)) {
```

File: cpp/oneapi/dal/detail/array_impl.hpp [hunk 7]
```diff
@@ -163,22 +162,47 @@ class array_impl : public base {
...
         }
         else {
-            data_owned_ = cshared(std::get<0>(ref.data_owned_), data);
+            if (const auto& ptr = std::get_if<0>(&ref.data_owned_)) {
+                data_owned_ = cshared(*ptr, data);
+            }
+            else {
+                throw internal_error(
+                    dal::detail::error_messages::array_does_not_contain_mutable_data());
+            }
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/detail/array_impl.hpp:205]: The same comments as for mutable version of `reset()`

File: cpp/oneapi/dal/detail/array_impl.hpp [hunk 8]
```diff
@@ -245,7 +269,7 @@ class array_impl : public base {
 #ifdef ONEDAL_DATA_PARALLEL
     std::optional<data_parallel_policy> dp_policy_;
 #endif
-};
+}; // namespace v1
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/detail/array_impl.hpp:272]:
```suggestion
};
```
It's a brace of `array_impl`

File: cpp/oneapi/dal/detail/array_impl.hpp (hunk 1 above)
Comment (PetrovKP) [reply cpp/oneapi/dal/detail/array_impl.hpp:95]: Done, thanks

File: cpp/oneapi/dal/detail/array_impl.hpp (hunk 5 above)
Comment (PetrovKP) [reply cpp/oneapi/dal/detail/array_impl.hpp:178]: applied

File: cpp/oneapi/dal/detail/array_impl.hpp (hunk 6 above)
Comment (PetrovKP) [reply cpp/oneapi/dal/detail/array_impl.hpp]: I only got it like this:
```c++
using shared_y = typename array_impl<Y>::shared;
if (const auto& ptr = std::get_if<shared_y>(&ref.data_owned_)) {
```

**Comment by PetrovKP** [comment]
/intelci: run
set daal4py_branch=dev/onedal4py
set daal4py_repo=git@github.com:PetrovKP/daal4py.git

**Comment by PetrovKP** [comment]
/intelci: run
set daal4py_branch=dev/onedal4py
set daal4py_repo=git@github.com:PetrovKP/daal4py.git

**Review by KulikovNikita — APPROVED** [review]

File: cpp/oneapi/dal/detail/error_messages.cpp [hunk 9]
```diff
@@ -26,6 +26,7 @@ namespace v1 {
 
 /* Common */
 MSG(array_does_not_contain_mutable_data, "Array does not contain mutable data")
+MSG(array_does_not_contain_ownership_structure, "Array does not contain ownership structure")
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/detail/error_messages.cpp:29]:
```suggestion
MSG(array_does_not_contain_ownership_structure, "Array is empty or does not contain ownership structure for the data")
```

**Review by michael-smirnov — APPROVED** [review]
Just a minor comment

**Pull request merged**